File: packages/native-android/android/src/main/java/com/openota/runtime/OpenOTALogger.kt
package com.openota.runtime

import android.util.Log

/**
 * Thin structured-logging wrapper. Kept as a single object (not a class per call site) so timing
 * spans read as one coherent trace in `adb logcat` without needing a full observability SDK wired
 * in yet — [timed] is the extension point a future analytics/crash-reporting hook would wrap.
 */
object OpenOTALogger {
    private const val TAG = "OpenOTA"

    @Volatile
    var debugMode: Boolean = false

    fun d(message: String) {
        if (debugMode) Log.d(TAG, message)
    }

    fun i(message: String) {
        Log.i(TAG, message)
    }

    fun w(message: String, throwable: Throwable? = null) {
        Log.w(TAG, message, throwable)
    }

    fun e(message: String, throwable: Throwable? = null) {
        Log.e(TAG, message, throwable)
    }

    /** Runs [block], logging its wall-clock duration under [label]. Rethrows on failure. */
    inline fun <T> timed(label: String, block: () -> T): T {
        val startedAt = System.nanoTime()
        try {
            val result = block()
            logDuration(label, startedAt, success = true)
            return result
        } catch (error: Throwable) {
            logDuration(label, startedAt, success = false)
            throw error
        }
    }

    fun logDuration(label: String, startedAtNanos: Long, success: Boolean) {
        val durationMs = (System.nanoTime() - startedAtNanos) / 1_000_000
        val status = if (success) "ok" else "failed"
        i("[$label] $status in ${durationMs}ms")
    }
}

File: packages/native-ios/ios/OpenOTALogger.swift
import Foundation
import os

/// Thin structured-logging wrapper mirroring `OpenOTALogger.kt`. Uses `os.Logger` (unified
/// logging) so spans are visible in Console.app / `xcrun simulator log` without pulling in an
/// observability SDK. `timed(_:_:)` is the extension point a future analytics hook would wrap.
enum OpenOTALogger {
    private static let logger = Logger(subsystem: "com.openota.runtime", category: "OpenOTA")

    static var debugMode: Bool = false

    static func d(_ message: String) {
        if debugMode { logger.debug("\(message, privacy: .public)") }
    }

    static func i(_ message: String) {
        logger.info("\(message, privacy: .public)")
    }

    static func w(_ message: String, error: Error? = nil) {
        if let error {
            logger.warning("\(message, privacy: .public): \(String(describing: error), privacy: .public)")
        } else {
            logger.warning("\(message, privacy: .public)")
        }
    }

    static func e(_ message: String, error: Error? = nil) {
        if let error {
            logger.error("\(message, privacy: .public): \(String(describing: error), privacy: .public)")
        } else {
            logger.error("\(message, privacy: .public)")
        }
    }

    /// Runs `block`, logging its wall-clock duration under `label`. Rethrows on failure.
    static func timed<T>(_ label: String, _ block: () throws -> T) rethrows -> T {
        let start = DispatchTime.now()
        do {
            let result = try block()
            logDuration(label, start: start, success: true)
            return result
        } catch {
            logDuration(label, start: start, success: false)
            throw error
        }
    }

    private static func logDuration(_ label: String, start: DispatchTime, success: Bool) {
        let ms = Double(DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000
        i("[\(label)] \(success ? "ok" : "failed") in \(String(format: "%.1f", ms))ms")
    }
}
